lib/bitcoin/BitcoinProcessor.ts

```diff
@@ -471,27 +493,37 @@ export default class BitcoinProcessor {
 
       const firstValidTransaction = await this.firstValidTransaction(exponentiallySpacedTransactions);
 
-      let revertToTransactionNumber: number;
-
       if (firstValidTransaction) {
-        // The number that represents the theoritical last possible transaction written on `firstValidTransaction.transactionTime`.
-        revertToTransactionNumber = TransactionNumber.construct(firstValidTransaction.transactionTime + 1, 0) - 1;
-      } else {
-        const lowestHeight = exponentiallySpacedTransactions[exponentiallySpacedTransactions.length - 1].transactionTime;
-        revertToTransactionNumber = TransactionNumber.construct(lowestHeight, 0);
-      }
+        // Revert all transactions in blocks from revertToBlockNumber and later. We make make this to be a group
+        // boundary to simplify resetting proof-of-fee state which is maintained per group.
+        let revertToBlockNumber = this.roundToGroupBoundary(firstValidTransaction.transactionTime);
 
-      console.debug(`Removing transactions since ${TransactionNumber.getBlockNumber(revertToTransactionNumber)}`);
-      await this.transactionStore.removeTransactionsLaterThan(revertToTransactionNumber);
+        // The number that represents the theoritical last possible transaction written with a block number
+        // less than revertToBlockNumber
+        const revertToTransactionNumber = TransactionNumber.construct(revertToBlockNumber + 1, 0) - 1;
 
-      // If we have reverted back to a valid transaction, rollback is complete.
-      if (firstValidTransaction) {
-        console.info(`reverted Transactions to block ${firstValidTransaction.transactionTime}`);
-        return firstValidTransaction.transactionTime;
+        console.debug(`Removing transactions since ${TransactionNumber.getBlockNumber(revertToTransactionNumber)}`);
+        await this.transactionStore.removeTransactionsLaterThan(revertToTransactionNumber);
+
+        // Reset transaction sampling
+        this.transactionSampler.clear();
+
+        // Revert the quantile calculator
+        await this.quantileCalculator.removeGroupsGreaterThanOrEqual(revertToBlockNumber + 1);
+
+        console.info(`reverted Transactions to block ${revertToBlockNumber}`);
+        return revertToBlockNumber;
       }
 
-      // Else we repeat the process using the newly reduced list of trasactions.
+      // We did not find a valid transaction - revert as much as the lowest height in the exponentially spaced
+      // transactions and repeat the process with a new reduced list of transactions.
+      const lowestHeight = exponentiallySpacedTransactions[exponentiallySpacedTransactions.length - 1].transactionTime;
+      const revertToTransactionNumber = TransactionNumber.construct(lowestHeight, 0);
+
+      console.debug(`Removing transactions since ${TransactionNumber.getBlockNumber(revertToTransactionNumber)}`);
+      await this.transactionStore.removeTransactionsLaterThan(revertToTransactionNumber);
     }
+
```

Comment:
Sanity: We would have a an infinite loop if the first transaction in DB is invalid.

Reply:
Opened issue #329 to track this issue.